Trying to build/test this against the bcm2711_defconfig I get the following errors
```
cc1: error: /home/kbingham/iob/libcamera/raspberrypi/sources/linux//home/kbingham/iob/libcamera/raspberrypi/sources/linux/drivers/media/pci/hailo: No such file or directory [-Werror=missing-include-dirs]
cc1: error: /home/kbingham/iob/libcamera/raspberrypi/sources/linux//home/kbingham/iob/libcamera/raspberrypi/sources/linux/drivers/media/pci/hailo/include: No such file or directory [-Werror=missing-include-dirs]
cc1: error: /home/kbingham/iob/libcamera/raspberrypi/sources/linux//home/kbingham/iob/libcamera/raspberrypi/sources/linux/drivers/media/pci/hailo/common: No such file or directory [-Werror=missing-include-dirs]
/home/kbingham/iob/libcamera/raspberrypi/sources/linux/drivers/media/pci/hailo/src/pcie.c:23:10: fatal error: hailo_pcie_version.h: No such file or directory
   23 | #include "hailo_pcie_version.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~
```

I don't think those are related to my branch here (Presumably caused in c5f657483e) so I'll disable HAILO and continue.

--- a/drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c
+++ b/drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c
@@ -3765,11 +3765,7 @@ static int bcm2835_codec_create(struct bcm2835_codec_driver *drv,
 
 	dev->role = role;
 
-#if 1
-	ret = -1;
-#else
 	ret = vchiq_mmal_init(&dev->instance);
-#endif
 	if (ret)
 		return ret;
 

--- a/drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c
+++ b/drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c
@@ -1713,11 +1713,7 @@ static int bcm2835_isp_probe_instance(struct platform_device *pdev,
 	if (ret)
 		return ret;
 
-#if 1
-	ret = -1;
-#else
 	ret = vchiq_mmal_init(&dev->mmal_instance);
-#endif
 	if (ret) {
 		v4l2_device_unregister(&dev->v4l2_dev);
 		return ret;

--- a/drivers/staging/vc04_services/interface/vchiq_arm/vchiq_arm.c
+++ b/drivers/staging/vc04_services/interface/vchiq_arm/vchiq_arm.c
@@ -804,7 +804,7 @@ int vchiq_initialise(struct vchiq_state *state, struct vchiq_instance **instance
 	 * block forever.
 	 */
 	for (i = 0; i < VCHIQ_INIT_RETRIES; i++) {
-		if (state)
+		if (vchiq_remote_initialised(state))
 			break;
 		usleep_range(500, 600);
 	}
@@ -1299,7 +1299,7 @@ void vchiq_dump_platform_instances(struct vchiq_state *state, struct seq_file *f
 {
 	int i;
 
-	if (!state)
+	if (!vchiq_remote_initialised(state))
 		return;
 
 	/*

--- a/drivers/staging/vc04_services/interface/vchiq_arm/vchiq_bus.h
+++ b/drivers/staging/vc04_services/interface/vchiq_arm/vchiq_bus.h
@@ -37,6 +37,16 @@ static inline struct vchiq_driver *to_vchiq_driver(struct device_driver *d)
 	return container_of(d, struct vchiq_driver, driver);
 }
 
+static inline void *vchiq_get_drvdata(const struct vchiq_device *device)
+{
+	return dev_get_drvdata(&device->dev);
+}
+
+static inline void vchiq_set_drvdata(struct vchiq_device *device, void *data)
+{
+	dev_set_drvdata(&device->dev, data);
+}
+
 extern const struct bus_type vchiq_bus_type;
 
 struct vchiq_device *

--- a/drivers/staging/vc04_services/interface/vchiq_arm/vchiq_core.h
+++ b/drivers/staging/vc04_services/interface/vchiq_arm/vchiq_core.h
@@ -413,6 +413,11 @@ struct vchiq_state {
 	struct opaque_platform_state *platform_state;
 };
 
+static inline bool vchiq_remote_initialised(const struct vchiq_state *state)
+{
+	return state->remote && state->remote->initialised;
+}
+
 struct bulk_waiter {
 	struct vchiq_bulk *bulk;
 	struct completion event;

--- a/drivers/staging/vc04_services/interface/vchiq_arm/vchiq_dev.c
+++ b/drivers/staging/vc04_services/interface/vchiq_arm/vchiq_dev.c
@@ -1170,6 +1170,11 @@ static int vchiq_open(struct inode *inode, struct file *file)
 
 	dev_dbg(state->dev, "arm: vchiq open\n");
 
+	if (!vchiq_remote_initialised(state)) {
+		dev_err(state->dev, "arm: vchiq has no connection to VideoCore\n");
+		return -ENOTCONN;
+	}
+
 	instance = kzalloc(sizeof(*instance), GFP_KERNEL);
 	if (!instance)
 		return -ENOMEM;
@@ -1200,7 +1205,7 @@ static int vchiq_release(struct inode *inode, struct file *file)
 
 	dev_dbg(state->dev, "arm: instance=%p\n", instance);
 
-	if (!state) {
+	if (!vchiq_remote_initialised(state)) {
 		ret = -EPERM;
 		goto out;
 	}

--- a/drivers/staging/vc04_services/vc-sm-cma/vc_sm.c
+++ b/drivers/staging/vc04_services/vc-sm-cma/vc_sm.c
@@ -48,6 +48,7 @@
 #include <linux/types.h>
 #include <asm/cacheflush.h>
 
+#include "vchiq_connected.h"
 #include "vc_sm_cma_vchi.h"
 
 #include "vc_sm.h"
@@ -1491,9 +1492,6 @@ static void vc_sm_connected_init(void)
 	 * Initialize and create a VCHI connection for the shared memory service
 	 * running on videocore.
 	 */
-#if 1
-	ret = -1;
-#else
 	ret = vchiq_initialise(&sm_state->vchiq_instance);
 	if (ret) {
 		pr_err("[%s]: failed to initialise VCHI instance (ret=%d)\n",
@@ -1501,7 +1499,7 @@ static void vc_sm_connected_init(void)
 
 		return;
 	}
-#endif
+
 	ret = vchiq_connect(sm_state->vchiq_instance);
 	if (ret) {
 		pr_err("[%s]: failed to connect VCHI instance (ret=%d)\n",
@@ -1592,9 +1590,7 @@ static int bcm2835_vc_sm_cma_probe(struct platform_device *pdev)
 	/* dma_set_max_seg_size checks if dma_parms is NULL. */
 	dma_set_max_seg_size(&pdev->dev, 0x3FFFFFFF);
 
-#if 0
 	vchiq_add_connected_callback(vc_sm_connected_init);
-#endif
 	return 0;
 }
 

